### DISCUSSION TITLE
Problem with PlcProgramBrowseAsync with Struct[]

### DISCUSSION THREAD

**Discussion by micheleBridi** [siemens/simatic-s7-webserver-api#74] [General]
I'm trying to make the browser of a datablock 42, its structure is this:
![image](https://github.com/siemens/simatic-s7-webserver-api/assets/74403751/42d9faea-be7a-4691-8b0e-8ec934dde79f)

But when I try to browse AlarmsTime of type struct[] I get this error:
The dimensions and edges of the array-indexes do not match the typeinformation of the plc. Did you provide a (correct) index?

the code I'm using is this:

`
 public async Task PlcProgramBrowse()
 {
     try
     {
        var allMyPlcProgramBlocks = (await requestHandler.PlcProgramBrowseAsync(ApiPlcProgramBrowseMode.Children)).Result;

         foreach (var block in allMyPlcProgramBlocks.Where(f => f.Db_number is 42))
             await RetrieveInfo(block);
     }
     catch (Exception e)
     {
         Console.WriteLine(e.Message);
     }
 }

 public async Task RetrieveInfo(ApiPlcProgramData block)
 {
     try
     {
         if (block.Has_children ?? false)
         {
             var items = (await requestHandler.PlcProgramBrowseAsync(ApiPlcProgramBrowseMode.Children, block)).Result;

             block.Children = items;

             foreach (var child in block.Children)
                 child.Parents = GetFather(block);

             foreach (var item in items)
             {
                 try
                 {
                     switch (item.Datatype)
                     {
                         case ApiPlcProgramDataType.Struct:
                         case ApiPlcProgramDataType.DataBlock:
                             await RetrieveInfo(item);
                             break;
                         default:                          
                                 Symbolics.Add(item);
break;
                     }
                 }
                 catch (Exception e)
                 {
                     Console.WriteLine(e.Message);
                 }
             }
         }
     }
     catch (Exception e)
     {
         Console.WriteLine(e.Message);
     }
 }

 public List<ApiPlcProgramData> GetFather(ApiPlcProgramData myPlcProgramDataBlock)
 {
     var result = new List<ApiPlcProgramData>();

     if (myPlcProgramDataBlock.Parents != null && myPlcProgramDataBlock.Parents.Count != 0)
         result.AddRange(myPlcProgramDataBlock.Parents);

     result.Add(myPlcProgramDataBlock);

     return result;
 }
`

**Comment by micheleBridi** [comment]
I managed to browse the entire datablock, the error was that it also passed the name of the array to the browse function.
Previously I passed it AlarmsTime.AlarmsTime.AlarmsTime[0] and now I pass it AlarmsTime.AlarmsTime[0].

However, the navigation function returns the two variables Start and End but their address is always 0 for start and 8 for end, both for AlarmsTime[0] and for AlarmsTime[1],AlarmsTime[2]AlarmsTime[3],...

**Reply by KircMax** [reply]
Hi @micheleBridi 
Sorry for being so unresponsive, I'm on vacation currently. I'll get back to you probably around the 15th 
Sorry again...

**Reply by KircMax** [reply]
But generally please check out the implementation in the PlcProgramHandler - I think I did some Struct and array Handling there, maybe it helps.

**Reply by micheleBridi** [reply]
Hi @KircMax 

i tried to use your method but the result is the same, i get all time the address of the first element of the array.

![image](https://github.com/siemens/simatic-s7-webserver-api/assets/74403751/797c4222-3984-4599-a9b6-03594e33f2c6)

**Reply by KircMax** [reply]
Hi @micheleBridi 
Ahh, now I can see what your issue is ... 
You'd expect to have the [1]."real_0" address to be like '%DB1%.DBD362' and the the [1]."real_1" address to be like '%DB1%.DBD366', right?

**Reply by KircMax** [reply]
what plc firmware version are you using?

**Reply by KircMax** [reply]
Nevermind. The issue is reproducible with the current firmware version. 
We have created a Bug for this... 
I'll update the issue when a fix is available:
https://github.com/siemens/simatic-s7-webserver-api/issues/75

**Reply by micheleBridi** [reply]
Hi @KircMax ,
when will the fix be available more or less?

**Reply by KircMax** [reply]
Hi @micheleBridi,
I cannot fix this on the library side...
The plc firmware will need to be updated and I cannot say, when the next release/update release will be, sorry. 
I just think for quite some time this issue will stay... sorry again.

**Reply by KircMax** [reply]
@micheleBridi 
The issue should be fixed with Firmware Release 3.1.4 - feel free to try it out and give feedback from your side :)
